sigstore-java/src/main/java/dev/sigstore/tuf/TufClient.java

```diff
@@ -0,0 +1,279 @@
+        throw new RoleVersionException(nextVersion, newRoot.getSignedMeta().getVersion());
+      }
+      // 5.3.7) set the trusted root metadata to the new root
+      trustedRoot = newRoot;
+      // 5.3.8) persist to repo
+      Path localTrustRoot = localStore.resolve("root.json");
+      if (localTrustRoot.toFile().exists()) {
+        // Backup the old root. (not sure if this is necessary)
+        Files.move(localTrustRoot, localStore.resolve((nextVersion - 1) + ".root.json"));
+      }
+      try (FileWriter fileWriter = new FileWriter(localTrustRoot.toFile())) {
+        fileWriter.write(GSON.get().toJson(trustedRoot));
+      }
+      // 5.3.9) see if there are more versions go back 5.3.3
+      nextVersion++;
+    }
+
+    // 5.3.10) Check expiration timestamp in trusted root is higher than fixed update start time,
+    // otherwise throw error.
+    ZonedDateTime expires = trustedRoot.getSignedMeta().getExpiresAsDate();
+    if (expires.isBefore(updateStartTime)) {
+      throw new RootExpiredException(mirror.toString(), updateStartTime, expires);
+    }
+    // 5.3.11) If the timestamp and / or snapshot keys have been rotated, then delete the trusted
+    // timestamp and snapshot metadata files.
+    if (hasNewKeys(preUpdateSnapshotRole, trustedRoot.getSignedMeta().getRole(Role.Name.SNAPSHOT))
+        || hasNewKeys(
+            preUpdateTimestampRole, trustedRoot.getSignedMeta().getRole(Role.Name.TIMESTAMP))) {
+      File snapshotMetaFile = localStore.resolve("snapshot.json").toFile();
+      if (snapshotMetaFile.exists()) {
+        Files.delete(snapshotMetaFile.toPath());
+      }
+      File timestampMetaFile = localStore.resolve("timestamp.json").toFile();
+      if (timestampMetaFile.exists()) {
+        Files.delete(timestampMetaFile.toPath());
+      }
+    }
+  }
+
+  private boolean hasNewKeys(RootRole oldRole, RootRole newRole) {
+    return newRole.getKeyids().stream().allMatch(s -> oldRole.getKeyids().contains(s));
+  }
+
+  /**
+   * Verifies that a delegate role has been signed by the threshold amount of keys.
+   *
+   * @param signatures these are the signatures on the role meta we're verifying
+   * @param threshold the minimum amount of signatures that must be verified by a trusted key.
+   * @param rootKeys a map of trusted keys used for signing, keyed by key_id
+   * @throws SignatureVerificationException if there are not enough verified signatures
+   */
+  private void verifyDelegate(
+      List<Signature> signatures,
+      int threshold,
+      Map<String, Key> rootKeys,
+      byte[] verificationMaterial)
+      throws SignatureVerificationException, NoSuchAlgorithmException, InvalidKeyException,
+          InvalidKeySpecException {
+    // use set to not count the same key multiple times towards the threshold.
+    var goodSigs = new HashSet<>(signatures.size());
+    for (Signature signature : signatures) {
+      var key = rootKeys.get(signature.getKeyId());
```

Comment:
shouldn't you need to check that this key ID is also signed off on that particular role?

e.g. we don't want to consider the timestmap key signing the snapshot role as valid, even though they are both key IDs in the root.

probably the code here should change:
```
 var newRootKeys = newRoot.getSignedMeta().getKeys(ROLE);
      // Verify our new root meta against the new root keys.
      verifyDelegate(
          newRootSignatures, newRootRoleMeta.getThreshold(), newRootKeys, newRootMetaBytes);
```